drv/riscv-plic/src/main.rs

```diff
@@ -0,0 +1,190 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+#![no_main]
+
+use riscv::plic;
+use riscv::plic::Plic;
+use userlib::*;
+
+use core::result::Result;
+
+use drv_riscv_plic_api::RiscvIntCtrlError;
+use idol_runtime::RequestError;
+use idol_runtime::RequestError::Runtime;
+
+const PLIC_IRQ: u32 = 0x0000_0001;
+
+fn get_irq_owner(irq: u32) -> Result<(TaskId, u32), ()> {
+    return match unsafe { HUBRIS_IRQ_TASK_LOOKUP.get(InterruptNum(irq)) } {
+        Some(task) => return Ok(*task),
+        None => Err(()),
+    };
+}
+
+fn set_irq_owner(irq: u32, owner: (TaskId, u32)) -> Result<(), ()> {
+    return unsafe { HUBRIS_IRQ_TASK_LOOKUP.set(InterruptNum(irq), owner) };
+}
+
+fn get_task_irqs(
+    owner: InterruptOwner,
+) -> Result<&'static [userlib::InterruptNum], ()> {
+    return match HUBRIS_TASK_IRQ_LOOKUP.get(owner) {
+        Some(irqs) => Ok(irqs),
+        None => Err(()),
+    };
+}
+
+fn irq_assigned(irq: u32) -> bool {
+    return unsafe { HUBRIS_IRQ_TASK_LOOKUP.contains(InterruptNum(irq)) };
+}
+
+#[repr(C)]
+struct ServerImpl {}
+
+impl idl::InOrderRiscvIntCtrlImpl for ServerImpl {
+    /// Disables the selected interrupt on the PLIC.
+    fn disable_int(
+        &mut self,
+        msg: &userlib::RecvMessage,
+        irq: u32,
+    ) -> Result<(), RequestError<RiscvIntCtrlError>> {
+        let owner: InterruptOwner = InterruptOwner {
+            task: msg.sender.index() as u32,
+            notification: irq,
+        };
+
+        match get_task_irqs(owner) {
+            Ok(irqs) => {
+                let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+                plic.mask(0, irqs[0].0.try_into().unwrap());
```

Comment:
Is the assumption here that a task can have only a single interrupt number associated with a notification? This seems to contradict how the interrupt number is defined as a vector and handled in build.rs: https://github.com/rivosinc/hubris/pull/7/commits/95a67ca0eb0d3d2fb07d3aa1c51b1078439a5b52#diff-0d889df1d35cb9c1a214e153b69a2bc77b2fb16bf1f678b5eb35c54956a310dfR112

Reply:
That's a mistake. Fixed.